File: library/src/main/java/nz/co/cic/mdns/MDNSAdvertiser.java
package nz.co.cic.mdns;

import android.content.Context;
import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;

import org.jdeferred.Deferred;

/**
 * Created by dipshit on 3/03/17.
 */

public class MDNSAdvertiser implements NsdManager.RegistrationListener {

    private Context mContext;
    private Deferred deferred;

    public MDNSAdvertiser(Context c){
        this.mContext = c;
    }

    public void advertise(NsdManager nsdManager, MDNSAdvertisement advertisement, Deferred deferred){
        this.deferred = deferred;
        NsdServiceInfo serviceInfo = new NsdServiceInfo();
        serviceInfo.setServiceName(advertisement.serviceName);
        serviceInfo.setPort(advertisement.servicePort);
        serviceInfo.setServiceType(advertisement.serviceType);
        nsdManager.registerService(serviceInfo, NsdManager.PROTOCOL_DNS_SD, this);
    }

    public void stop(NsdManager nsdManager, Deferred deferred){
        this.deferred = deferred;
        nsdManager.unregisterService(this);
    }

    @Override
    public void onRegistrationFailed(NsdServiceInfo nsdServiceInfo, int i) {
        this.deferred.reject(i);
    }

    @Override
    public void onUnregistrationFailed(NsdServiceInfo nsdServiceInfo, int i) {
        this.deferred.reject(i);
    }

    @Override
    public void onServiceRegistered(NsdServiceInfo nsdServiceInfo) {
        this.deferred.resolve(nsdServiceInfo);
    }

    @Override
    public void onServiceUnregistered(NsdServiceInfo nsdServiceInfo) {
        this.deferred.resolve(nsdServiceInfo);
    }
}

File: library/src/main/java/nz/co/cic/mdns/MDNSAdvertiser.kt
package nz.co.cic.mdns

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableEmitter

/**
 * Created by dipshit on 3/03/17.
 */

class MDNSAdvertiser(private val mContext: Context) : NsdManager.RegistrationListener {

    private var nsdManager : NsdManager? = null
    private var observer: FlowableEmitter<MDNSStatus>? = null
    private var dObserver: FlowableEmitter<MDNSStatus>? = null

    init {
        this.nsdManager = mContext.getSystemService(Context.NSD_SERVICE) as NsdManager
    }

    fun advertise(advertisement: MDNSAdvertisement): Flowable<MDNSStatus> {
        return Flowable.create({
            subscriber ->
            this.observer = subscriber
            val serviceInfo = NsdServiceInfo()
            serviceInfo.serviceName = advertisement.serviceName
            serviceInfo.port = advertisement.servicePort
            serviceInfo.serviceType = advertisement.serviceType
            nsdManager?.registerService(serviceInfo, NsdManager.PROTOCOL_DNS_SD, this)
        }, BackpressureStrategy.BUFFER)
    }

    fun stop(nsdManager: NsdManager): Flowable<MDNSStatus> {
        return Flowable.create({
            subscriber ->
            this.dObserver = subscriber
            nsdManager.unregisterService(this)
        }, BackpressureStrategy.BUFFER)
    }

    override fun onRegistrationFailed(nsdServiceInfo: NsdServiceInfo, i: Int) {
        this.observer?.onError(Throwable("Failed to register"))
    }

    override fun onUnregistrationFailed(nsdServiceInfo: NsdServiceInfo, i: Int) {
        this.dObserver?.onError(Throwable("Failed to unregister"))
    }

    override fun onServiceRegistered(nsdServiceInfo: NsdServiceInfo) {
        observer?.onNext(MDNSStatus("Service Registered"))
        this.observer?.onComplete()
    }

    override fun onServiceUnregistered(nsdServiceInfo: NsdServiceInfo) {
        observer?.onNext(MDNSStatus("Service Unregistered"))
        this.dObserver?.onComplete()
    }
}
